File: deck/seed/code/native/kotlin/file/asynchronous/runtime/stat.kt
// File metadata for the kotlin target, over java.nio.file's attribute views, read on `Dispatchers.IO`. It builds
// the emitted `FileMetadata` record directly (the shim is prepended to the module that declares it, so the type is
// in scope), which is what keeps one attribute read from becoming seven. Reached only through the public
// file/metadata API.
import java.nio.file.Files
import java.nio.file.LinkOption
import java.nio.file.Paths
import java.nio.file.attribute.PosixFileAttributes
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object stat {
  suspend fun metaRead(path: String, follow: Boolean): FileMetadata =
    withContext(Dispatchers.IO) {
      try {
        val at = Paths.get(path)
        val options =
          if (follow) emptyArray<LinkOption>() else arrayOf(LinkOption.NOFOLLOW_LINKS)
        val info = Files.readAttributes(at, PosixFileAttributes::class.java, *options)

        val kind =
          when {
            info.isDirectory -> "directory"
            info.isSymbolicLink -> "link"
            info.isRegularFile -> "file"
            else -> "other"
          }

        var mode = 0L

        for (one in info.permissions()) {
          mode = mode or
            when (one) {
              java.nio.file.attribute.PosixFilePermission.OWNER_READ -> 0x100L
              java.nio.file.attribute.PosixFilePermission.OWNER_WRITE -> 0x80L
              java.nio.file.attribute.PosixFilePermission.OWNER_EXECUTE -> 0x40L
              java.nio.file.attribute.PosixFilePermission.GROUP_READ -> 0x20L
              java.nio.file.attribute.PosixFilePermission.GROUP_WRITE -> 0x10L
              java.nio.file.attribute.PosixFilePermission.GROUP_EXECUTE -> 0x8L
              java.nio.file.attribute.PosixFilePermission.OTHERS_READ -> 0x4L
              java.nio.file.attribute.PosixFilePermission.OTHERS_WRITE -> 0x2L
              java.nio.file.attribute.PosixFilePermission.OTHERS_EXECUTE -> 0x1L
            }
        }

        FileMetadata(
          info.size(),
          kind,
          info.creationTime().toMillis(),
          info.lastModifiedTime().toMillis(),
          info.lastAccessTime().toMillis(),
          mode,
          info.isSymbolicLink,
        )
      } catch (error: Exception) {
        // a missing path reads as the zero record rather than throwing: the public API is total
        FileMetadata(0L, "other", 0L, 0L, 0L, 0L, false)
      }
    }
}

File: deck/seed/code/native/rust/file/asynchronous/runtime/stat.rs
// File metadata for the rust target, over `tokio::fs::metadata`. It builds the emitted `FileMetadata` record
// directly (the shim is prepended to the module that declares it, so the type is in scope through `use super::*`),
// which is what keeps one `stat` call from becoming seven. Reached only through the public file/metadata API.
mod stat {
    use super::FileMetadata;
    use std::os::unix::fs::MetadataExt;
    use std::os::unix::fs::PermissionsExt;

    // milliseconds since the epoch, the unit every backend's metadata reports, and 0 for a filesystem that does
    // not record the field (birth time on most Linux filesystems)
    fn moment(time: std::io::Result<std::time::SystemTime>) -> i64 {
        time.ok()
            .and_then(|at| at.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|since| since.as_millis() as i64)
            .unwrap_or(0)
    }

    pub async fn meta_read(path: String, follow: bool) -> FileMetadata {
        let meta = if follow {
            tokio::fs::metadata(&path).await
        } else {
            tokio::fs::symlink_metadata(&path).await
        };

        let meta = match meta {
            Ok(meta) => meta,
            // a missing path reads as the zero record rather than unwinding: the public API is total
            Err(_) => {
                return FileMetadata {
                    size: 0,
                    kind: "other".to_string(),
                    made: 0,
                    changed: 0,
                    opened: 0,
                    mode: 0,
                    link: false,
                }
            }
        };

        let kind = if meta.is_dir() {
            "directory"
        } else if meta.file_type().is_symlink() {
            "link"
        } else if meta.is_file() {
            "file"
        } else {
            "other"
        };

        FileMetadata {
            size: meta.len() as i64,
            kind: kind.to_string(),
            made: moment(meta.created()),
            changed: meta.mtime() * 1000,
            opened: meta.atime() * 1000,
            mode: meta.permissions().mode() as i64,
            link: meta.file_type().is_symlink(),
        }
    }
}
